pkg/query-service/app/queryBuilder/functions.go

```diff
@@ -0,0 +1,242 @@
+)
+
+// funcCutOffMin cuts off values below the threshold and replaces them with NaN
+func funcCutOffMin(result *v3.Result, threshold float64) *v3.Result {
+	for _, series := range result.Series {
+		for idx, point := range series.Points {
+			if point.Value < threshold {
+				point.Value = math.NaN()
+			}
+			series.Points[idx] = point
+		}
+	}
+	return result
+}
+
+// funcCutOffMax cuts off values above the threshold and replaces them with NaN
+func funcCutOffMax(result *v3.Result, threshold float64) *v3.Result {
+	for _, series := range result.Series {
+		for idx, point := range series.Points {
+			if point.Value > threshold {
+				point.Value = math.NaN()
+			}
+			series.Points[idx] = point
+		}
+	}
+	return result
+}
+
+// funcClampMin cuts off values below the threshold and replaces them with the threshold
+func funcClampMin(result *v3.Result, threshold float64) *v3.Result {
+	for _, series := range result.Series {
+		for idx, point := range series.Points {
+			if point.Value < threshold {
+				point.Value = threshold
+			}
+			series.Points[idx] = point
+		}
+	}
+	return result
+}
+
+// funcClampMax cuts off values above the threshold and replaces them with the threshold
+func funcClampMax(result *v3.Result, threshold float64) *v3.Result {
+	for _, series := range result.Series {
+		for idx, point := range series.Points {
+			if point.Value > threshold {
+				point.Value = threshold
+			}
+			series.Points[idx] = point
+		}
+	}
+	return result
+}
+
+// funcAbsolute returns the absolute value of each point
+func funcAbsolute(result *v3.Result) *v3.Result {
+	for _, series := range result.Series {
+		for idx, point := range series.Points {
+			point.Value = math.Abs(point.Value)
+			series.Points[idx] = point
+		}
+	}
+	return result
+}
+
+// funcLog2 returns the log2 of each point
+func funcLog2(result *v3.Result) *v3.Result {
+	for _, series := range result.Series {
+		for idx, point := range series.Points {
+			point.Value = math.Log2(point.Value)
+			series.Points[idx] = point
+		}
+	}
+	return result
+}
+
+// funcLog10 returns the log10 of each point
+func funcLog10(result *v3.Result) *v3.Result {
+	for _, series := range result.Series {
+		for idx, point := range series.Points {
+			point.Value = math.Log10(point.Value)
+			series.Points[idx] = point
+		}
+	}
+	return result
+}
+
+// funcCumSum returns the cumulative sum for each point in a series
+func funcCumSum(result *v3.Result) *v3.Result {
+	for _, series := range result.Series {
+		var sum float64
+		for idx, point := range series.Points {
+			sum += point.Value
+			point.Value = sum
+			series.Points[idx] = point
+		}
+	}
+	return result
+}
+
+// funcEWMA3 returns the exponential weighted moving average for each point in a series
+func funcEWMA3(result *v3.Result, alpha float64) *v3.Result {
+	for _, series := range result.Series {
+		var prev float64
+		for _, point := range series.Points {
+			if math.IsNaN(prev) {
+				point.Value = point.Value
```

Comment:
Do you think cases when `point.Value` can be NaN as functions can be chained. Eg say `funcEWMA3 (clampMax())`?

Reply:
Yes, it's possible but it's a good point. Let me think about if the same should be considered in other functions such as cumulative sum.

Reply:
Handled NaN in all relevenat functions